Refs #340: Swap old 'CLA' for the DCO

GregSutcliffe opened the pull request:
Per #340 this is the first part of the process to swap out the old CLA for the DCO. It:

- Adds the CONTRIBUTING and DCO files to the project root
  - this is the recommended way of informing people about the DCO
- removes the old CLA file
- replaces it with a new DCO doc that has the same content as CONTRIBUTING (and thus links to the DCO file in the root)
- Updates a couple of other files that mention the CLA to cover the DCO instead/as well

This is only the first part, we'll need the two root files (CONTRIBUTING & DCO) in other repos too.

I'm marking this as draft while we discuss if I missed anything :stuck_out_tongue:

GregSutcliffe committed:
Refs #340: Swap old 'CLA' for the DCO

gotmax23 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,11 @@
+# How to Contribute
+
+This project is [GPL-3.0 licensed](COPYING) and accepts contributions via
```
I'd prefer to be more explicit here.

```suggestion
The ansible-documention project is [GPL-3.0 licensed](COPYING) and accepts contributions via
```

GregSutcliffe replied:
LGTM, thanks!

gotmax23 commented on docs/docsite/rst/community/collection_contributors/collection_requirements.rst:
```diff
@@ -321,8 +321,8 @@ These guidelines are the policy for inclusion in the Ansible package and are in
 
-Contributor License Agreements
-==============================
+Contributor License Agreements (CLAs) and the DCO
```
I don't think this type of thing is appropriate for the Collection Requirements. It focuses more on what the guidelines are and less on how to follow them.

GregSutcliffe replied:
Fair enough. We can move this part to a later stage, perhaps add it to the inclusion criteria? @samccann is working on a project template that will include the DCO too, so that will help. I can remove this for now, I think.

samccann replied:
This section was likely added by @Andersson007 so just giving him a ping in case he has opinions.

gotmax23 replied:
I would remove this from here and add the DCO files to https://github.com/ansible-collections/collection_template/ instead. All new collections are created from that template.

GregSutcliffe replied:
If we move it, lets move all of it. It doesn;t make sense to mention the prohibition of CLAs without mentioning the DCO. It's all or nothing, but I'm OK with nothing if it moves to the template.

gotmax23 replied:
> If we move it, lets move all of it. It doesn;t make sense to mention the prohibition of CLAs without mentioning the DCO. It's all or nothing, but I'm OK with nothing if it moves to the template.


The Collection Requirements already mention the DCO. The template repo gives an example repo structure that follows the Collection Requirements, but it is not a place to encode requirements themselves. Adding the example DCO and CONTRIBUTING.md files there would make sense if we want new collections to pick them up.


If we want to _recommend_ that collections adopt the DCO in the Collecrtion Requirements, that can be brought to the SC for consideration, but adding a link to some other repo that the SC doesn't own as _an example_ showing _how_ to adopt the DCO without extra context is out of scope. In a similar vain, we (up until recently) required that collections still support Python 2.7 alongside Python 3, but we didn't explain that collections need to be careful of str/bytes or recommend using the six library.

GregSutcliffe replied:
Your point about scope is well made, so I'll remove this for now. However, I do want to think about how we aim for consistency across the ecosystem - I'll follow that up with an SC forum topic once we've got this done.

gotmax23 commented:
Is ansible-core also adopting the DCO?
I'd lean towards dropping the current useless CLA in one step and adding the DCO in another.

felixfontein commented on docs/docsite/rst/dev_guide/ansible_index.rst:
```diff
@@ -59,7 +59,7 @@ Find the task that best describes what you want to do:
   * I want to :ref:`understand how to contribute to Ansible <ansible_community_guide>`.
   * I want to :ref:`contribute my module or plugin <developing_modules_checklist>`.
-  * I want to :ref:`understand the license agreement <contributor_license_agreement>` for contributions to Ansible.
+  * I want to :ref:`understand the license agreement <developer_certificate_of_origin>` for contributions to Ansible.
```
Shouldn't here `license` be changed to `DCO` as well (as in docs/docsite/rst/dev_guide/developing_modules_checklist.rst)? Same for the next ones.

Or add `(DCO)` in brackets, so that folks know that this isn't some CLA?

GregSutcliffe replied:
Good catch, fixed

GregSutcliffe commented:
> Is ansible-core also adopting the DCO? I'd lean towards dropping the current useless CLA in one step and adding the DCO in another.

What's your thinking here @gotmax23? Given the proposal is to replace one with the other, is there much advantage to be gained in splitting the commits up? Do we want even a short period of time when **no** agreement is present in the repo?

gotmax23 commented:
> > Is ansible-core also adopting the DCO? I'd lean towards dropping the current useless CLA in one step and adding the DCO in another.
> 
> What's your thinking here @gotmax23? Given the proposal is to replace one with the other, is there much advantage to be gained in splitting the commits up? Do we want even a short period of time when **no** agreement is present in the repo?

I'm fine with doing them at once. I just wasn't sure coordinating with core would be more of an ordeal and would prefer to remove the current "broken" CLA sooner rather than later.

gotmax23 commented:
(We're discussing this in the DAWGs meeting. I'll edit this with notes later.)

- We discussed adding a checkbox to PRs or requiring Signed-off-by on PR. According to (@GregSutcliffe speaking on behalf of) Legal, it's okay to simply have the DCO file in the repo and documenting in CONTRIBUTING as this PR does without requiring additional attestations.
- I want to clarify which repos this applies to. Saying the DCO applies to contributions to `Ansible` is ambiguous.

---

From the [meeting log](https://forum.ansible.com/t/meeting-log-ansible-docs-2023-10-17-1504/1756):

- INFO: for https://github.com/ansible/ansible-documentation/pull/598 - implementing DCO - does it have to be timed with ansible/ansible as well? (@samccann:ansible.im, 15:25:59)
- INFO: most DCO repos also do signed commits but that seems excessive for docs, so we can perhaps add a checkbox to the PR template so contributors have to check it to ssay they agree with the DCO (@samccann:ansible.im, 15:26:58)
- INFO: CLI/DCO PR wording suggests DCO applies to all Ansible (aka overloaded term) so should be specific to core and collections in the dev guide updates (with core approval) (@samccann:ansible.im, 15:42:54)

gotmax23 commented on docs/docsite/rst/dev_guide/core_index.rst:
```diff
@@ -56,7 +56,7 @@ Find the task that best describes what you want to do:
   * I want to :ref:`understand how to contribute to Ansible <ansible_community_guide>`.
   * I want to :ref:`contribute my module or plugin <developing_modules_checklist>`.
-  * I want to :ref:`understand the license agreement <contributor_license_agreement>` for contributions to Ansible.
+  * I want to :ref:`understand the DCO agreement <developer_certificate_of_origin>` for contributions to Ansible.
```
Assuming core agrees to this, I'd change this text and the rest of the places it is repeated to something like

```suggestion
  * I want to :ref:`understand the DCO agreement <developer_certificate_of_origin>` for contributions to the `Ansible Core <https://github.com/ansible/ansible>`_ and `Ansible Documention <https://github.com/ansible/ansible-documentation>`_ repositories.
```

GregSutcliffe replied:
OK, we could use this for now. However, I'd like to think about how that's going to read when we're approaching 40%, 50%, 60% of the project signed up to the DCO. We'll end up with frequent updates here and a *very* long line - so we need to think about when it's acceptable to say "Ansible".

As an alternative, consider that we could use wording like "I want to understand the DCO agreement for contributions to Ansible projects (where applicable, see the CONTRIBUTING in project repo)" which then gives a carve out for those who haven't got there yet, and needs less updating - how does that feel?

samccann replied:
Something to keep in mind - these docs are only applicable to ansible-core and collections. Dev guides for every other Ansible ecosystem project is handled by those projects, not this doc.

GregSutcliffe replied:
Ah OK, that helps. In that case I'm OK with the change if you are @samccann?

samccann replied:
yep sounds good to me

gotmax23 replied:
> Ah OK, that helps. In that case I'm OK with the change if you are @samccann?


To be clear, does `the change` refers to my suggestion or @GregSutcliffe's?

GregSutcliffe replied:
I'm saying I'm happy with your proposed alteration @gotmax23.

gotmax23 commented on docs/docsite/rst/dev_guide/developing_modules_checklist.rst:
```diff
@@ -20,7 +20,7 @@ To contribute a module to most Ansible collections, you must:
 * follow `PEP 8 <https://www.python.org/dev/peps/pep-0008/>`_ Python style conventions - see :ref:`testing_pep8` for more information
 * license your module under the GPL license (GPLv3 or later)
-* understand the :ref:`license agreement <contributor_license_agreement>`, which applies to all contributions
+* understand the :ref:`DCO agreement <developer_certificate_of_origin>`, which applies to all contributions
```
```suggestion
* understand the :ref:`DCO agreement <developer_certificate_of_origin>`, which applies to contributions to the `Ansible Core <https://github.com/ansible/ansible>`_ and `Ansible Documention <https://github.com/ansible/ansible-documentation>`_ repositories.
```

GregSutcliffe committed:
Update CONTRIBUTING.md

Co-authored-by: Maxwell G <maxwell@gtmx.me>

samccann commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,11 @@
+# How to Contribute
+
+The ansible-documention project is [GPL-3.0 licensed](COPYING) and accepts contributions via
```
```suggestion
The ansible-documention project is [GPL-3.0 licensed](COPYING) and accepts contributions through
```

GregSutcliffe committed:
Update CONTRIBUTING.md

Co-authored-by: Sandra McCann <samccann@redhat.com>

GregSutcliffe commented:
> I'm fine with doing them at once. I just wasn't sure coordinating with core would be more of an ordeal and would prefer to remove the current "broken" CLA sooner rather than later.

Ah I see. I think we need agreement for both steps, we can't drop the CLA unilaterally either, so it may as well stay as one PR?

samccann commented on docs/docsite/rst/community/collection_contributors/collection_requirements.rst:
```diff
@@ -321,8 +321,8 @@ These guidelines are the policy for inclusion in the Ansible package and are in
 
-Contributor License Agreements
-==============================
+Contributor License Agreements (CLAs) and the DCO
```
```suggestion
Contributor License Agreement (CLA) and the DCO
```

samccann replied:
My suggested edit above is only if we leave this here... it gets around the false spelling error CI failure.

GregSutcliffe replied:
I've removed CLAs for now so we can park this until it returns (if ever)

gotmax23 commented:
Retriggering CI

GregSutcliffe committed:
Update docs/docsite/rst/dev_guide/core_index.rst

Co-authored-by: Maxwell G <maxwell@gtmx.me>

GregSutcliffe committed:
Update docs/docsite/rst/dev_guide/developing_modules_checklist.rst

Co-authored-by: Maxwell G <maxwell@gtmx.me>

GregSutcliffe committed:
Remove DCO from collection_requirements.rst, and spelling fixes

GregSutcliffe commented:
I've committed the suggested changes, fixed the spelling errors, and removed the Collections part as discussed. I'm doing a matching PR to Core shortly, will link back from there too. @gotmax23 @samccann @richardfontana any concerns with this as it stands?

gotmax23 commented on docs/docsite/rst/dev_guide/ansible_index.rst:
```diff
@@ -59,7 +59,7 @@ Find the task that best describes what you want to do:
   * I want to :ref:`understand how to contribute to Ansible <ansible_community_guide>`.
   * I want to :ref:`contribute my module or plugin <developing_modules_checklist>`.
-  * I want to :ref:`understand the license agreement <contributor_license_agreement>` for contributions to Ansible.
+  * I want to :ref:`understand the DCO agreement <developer_certificate_of_origin>` for contributions to Ansible.
```
```suggestion
  * I want to :ref:`understand the DCO agreement <developer_certificate_of_origin>` for contributions to the `Ansible Core <https://github.com/ansible/ansible>`_ and `Ansible Documentation <https://github.com/ansible/ansible-documentation>`_ repositories.
```

felixfontein replied:
:+1: for this change.

gotmax23 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,11 @@
+
+## Certificate of Origin
+
+By contributing to this project you agree to the Developer Certificate of
```
```suggestion
By contributing to ansible-documentation, you agree to the Developer Certificate of
```

tvo318 replied:
Do you see benefit for other projects to reference this? If so, we should keep it generic?

gotmax23 replied:
Other projects can copy this into their own repository and change the project name.

gotmax23 reviewed (commented):
Other than those last two wording suggestions, the changes look good to me. Thanks for taking care of this!

felixfontein commented:
Is there a reason not to also add @gotmax23's last two suggestions?

GregSutcliffe committed:
Update CONTRIBUTING.md

Co-authored-by: Maxwell G <maxwell@gtmx.me>

GregSutcliffe committed:
Update docs/docsite/rst/dev_guide/ansible_index.rst

Co-authored-by: Maxwell G <maxwell@gtmx.me>

GregSutcliffe commented:
Apologies, I missed those somehow - accepted both. Hopefully the core one will be merged soon too!

Andersson007 commented:
as https://github.com/ansible/ansible/pull/82105 has been merged, can this one be also merged?
